cloudstate/cloudstate.go

```diff
@@ -0,0 +1,249 @@
+//
+// Copyright 2019 Lightbend Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cloudstateio/go-support/cloudstate/protocol"
+	"github.com/golang/protobuf/descriptor"
+	"github.com/golang/protobuf/proto"
+	filedescr "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+	"log"
+	"net"
+	"os"
+	"runtime"
+)
+
+const (
+	SupportLibraryVersion = "0.1.0"
+	SupportLibraryName    = "cloudstate-go-support"
+)
+
+// CloudState is an instance of a CloudState User Function
+type CloudState struct {
+	server                   *grpc.Server
+	entityDiscoveryResponder *EntityDiscoveryResponder
+	eventSourcedHandler      *EventSourcedHandler
+}
+
+// NewCloudState returns a new CloudState instance.
+func NewCloudState(options *Options) *CloudState {
+	cs := &CloudState{
+		server:                   grpc.NewServer(),
+		entityDiscoveryResponder: NewEntityDiscoveryResponder(options),
+		eventSourcedHandler:      NewEventSourcedHandler(),
+	}
+	protocol.RegisterEntityDiscoveryServer(cs.server, cs.entityDiscoveryResponder)
+	log.Println("RegisterEntityDiscoveryServer")
+	protocol.RegisterEventSourcedServer(cs.server, cs.eventSourcedHandler)
+	log.Println("RegisterEventSourcedServer")
```

Comment:
@jsravn great point. I removed logging considerably throughout revisions of the code, having a lot of logs in the beginning. I think we actually could remove logging completely or solely have debugging information if needed.
Your also stated in your main review comment, the user could also get some "state" of the userfunction support library.

As there are two aspects of the "logging problem" I think: 1. how to log in this Go based library and 2. how to log in general in a "userfunction support library" in CloudState. I brought the second question today in the Contributors Call and Viktor openend [an Issue to discuss it](https://github.com/cloudstateio/cloudstate/issues/130)

For part 1: what would you suggest. In the Go projects I did so far I logged without any 3rd party library so I'm open to any suitable option here that makes sense.

Reply:
I would say the most common loggers are [logrus](https://github.com/sirupsen/logrus) which provides structured logging (very handy), and glog (https://github.com/golang/glog).

Both have very similar interfaces - such that we could probably define a `Logger` interface in cloudstate that users can provide an implementation for. With us just defaulting to something if a user doesn't specify:

```
package cloudstate

type Logger interface {
  Info(args ...interface{})
  Infof(format string, args ...interface{})
  // and so on
}
```

You could also argue that this library is more of a framework for user functions. In which case it might be okay for us to be opinionated about logging, and expect the user function to use whatever logging we mandate. In that case I would lean towards logrus but that is my own bias :).

Reply:
If we have an interface one could argue that the implementation could be a non-invasive as possible, like the std-libs logging, so a user himself can choose whatever he wants to use.

I tend to have perhaps no logging at all and let the user choose a logger of he prefers. The point could also be clarified with https://github.com/cloudstateio/cloudstate/issues/130 so that user libraries behave similar in this regard.

Reply:
see https://github.com/cloudstateio/go-support/issues/13